explainer.md

```diff
@@ -215,11 +224,189 @@ defaults.
+If you know your input will contain elaborate markup and you only wish to
+retain some of it, but you want to retain the full textual content, you
+can use `dropElement`. This will remove only elements but retain their children.
+There is no `dropAttribute` because attribute nodes do not have children.
+
+```js
+const config_that_drops_things = {
+  dropElements: ["span", "em", "u", "s", "i", "b"]
+};
+```
+
+How do multiple element list types interact? In order for an element to
+pass, it must pass _all_ lists.
+
+You can use allow-lists and block-lists together, but note that this doesn't
+usually make much sense since any combination of allow- and block-lists can
+be rewritten as an allow-lists of equal or shorter length by removing any
+block-list element from the allow-list, or dropping it altogether if there
+was no matching allow-list entry in the first place.
+
+```js
+// equivalent to:  { allowElements: ["i"] }
+const config_that_mixes_allow_and_block_lists = {
+    allowElements: ["i", "u"],
+    blockElements: ["u", "s"],
+};
+
+// <div><i>A</i></div>
+element.setHTML("<i>A</i><u>B</u><s>C</s><b>D</b>",
+                { sanitizer: config_that_mixes_allow_and_block_lists });
+```
+
+A more useful combination of different list types would be a config that
+allows simple text structure and styling, but retains the full text:
+
+```js
+const config_that_retains_simple_styling_but_most_text = {
+  allowElements: ["p", "b", "i"],
+  dropElements: ["div", "span", "em", "u", "s", "li"],
+};
+const styled_text = "<p>Some <span style='color: blue'>colourful</span> <u>styled</u> <b>text</b>";
+
+// <div><p>Some colourful styled <b>text</b></p></div>
+element.setHTML(styled_text, {sanitizer: config_that_retains_simple_styling_but_most_text});
+```
+
+### Configuring attributes per element
+
+A common use case is to allow or block all instances of a given attribute,
+but this isn't always sufficient. One may also act on attributes on specific
+elements.
+
+In the example `config_text_with_only_simple_styling` in the previous chapter
+we have allowed the `src` attribute on any of the eight allowed elements.
+If one wanted to allow `class` everywhere, but `src` only on `<img>`, the
+following would do:
+
+```js
+const config_with_element_specific_attributes = {
+  allowElements: [
+    "p", "b", "i",
+    { name: "img", attributes: [ "src" ] }
+  ],
+  allowAttributes: ["class"],
+};
+```
+
+If you want to drop `src` attributes from `<input>` elements but retain them
+elsewhere, you can use:
+
+```js
+const drop_src_attribute_from_input = {
+  dropElements: [{ name: "input", attributes: ["src"]}],
+}
+```
+
+TODO: What does `{ name: "bla", attributes: [] }` mean? My gut feeling would
+  be that `attributes: []` is a no-op and it allows `<bla>` plus whatever
+  attributes are allowed globally. But if so, the only way to describe an
+  element without any attributes in presence of a global `allowAttributes`
+  would be to add a dropElements for it.
```

Comment:
To be clear, this question is for `allowAttributes: { name: "bla", attributes: [] }` right?
Hm, what if empty actually meant "no attributes for this element"? That would make the list look-up "just work" and allow special-casing elements where the global `allowAttributes` shouldn't apply (I am assuming both rules must allow the attribute).

Reply:
I meant `allowElements`, not `allowAttributes`.

Given the input `"<img src=bla srcset=blubb alt=blerk>"`, what do these do:

* `allowElements: [{ name: "img", attributes: [] }]`
* `allowElements: ["img"]`
* `allowElements: `[{name: "img", attributes: undefined }]`

My gut feeling is that these three should mean the same thing. (I could certainly be convinced otherwise.)

On: _"allow special-casing elements where the global `allowAttributes` shouldn't apply (I am assuming both rules must allow the attribute)."_

* Would `allowElements: [{ name: "img", attributes: [] }], allowAttributes: ["src", "srcset"]` turn the input example above into "<img>", because the attributes are not allowed by both rules?

I'm worried that we're making the "grab bag" of elements and attributes very laborious, which was the primary use case we found when looking at existing sanitizer usage.

Reply:
1. Yes, I think all of those three examples should resulti n `<img>` with no attributes.
2. Great point. We shouldn't make this too cumbersome, I agree. I guess this could work if our step for "finding the allowed attributes for this current element" were to build the union of what `allowElement` and `allowAttributes` are specifying. But then, how would you build an instruction that is "_allow an element `img` with no attributes at all, not even the ones specified elsewhere_". Can't think of something but happy to go with a suggestion from you.